src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/DefaultConnectionPlugin.java

```diff
@@ -100,8 +107,24 @@ public void openInitialConnection(ConnectionUrl connectionUrl) throws SQLExcepti
       return;
     }
 
-    HostInfo mainHostInfo = connectionUrl.getMainHost();
+    final HostInfo mainHostInfo = connectionUrl.getMainHost();
     JdbcConnection connection = this.connectionProvider.connect(mainHostInfo);
+
+    final RdsUtils rdsUtils = new RdsUtils();
+    if (rdsUtils.isReaderClusterDns(mainHostInfo.getHost())) {
+      final String connectedHostName = getCurrentlyConnectedInstance(connection);
+      String instanceEndpoint =
+          rdsUtils.getRdsInstanceHostPattern(mainHostInfo.getHost()).replace("?", connectedHostName);
+      if (!StringUtils.isNullOrEmpty(mainHostInfo.getHostProperties().get(PropertyKey.clusterInstanceHostPattern.getKeyName()))) {
+        instanceEndpoint = mainHostInfo.getHostProperties().get(PropertyKey.clusterInstanceHostPattern.getKeyName())
+                .replace("?", connectedHostName);
+      }
+      final HostInfo instanceHostInfo = ConnectionUtils.createInstanceHostWithProperties(instanceEndpoint, mainHostInfo);
+      final JdbcConnection hostConnection = this.connectionProvider.connect(instanceHostInfo);
+      connection.close();
+      connection = hostConnection;
+    }
```

Comment:
I'd suggest to relocate this logic into a new plugin. In this way, it'd be easier for users to use it when it's needed, or skip it. That gives more flexibility for users who don't experience any issue with query timeout.